Decouple Modular system from Asp.Net MVC

Jetski5822 opened the pull request:
# ASP.Net Modules

Modules provides a mechinism to have a self contained modular system where you can opt in to a specific application framework and not have the design of you application be dictated to by such.

This directly allows you to decouple the Hosting application from the independent modules. So no more redeploying the whole application for one simple change, instead you can just deploy a single module and recycle you application.

## Getting started

First, create a brand new web application.

Within this new application we are initially going to focus on two files, `project.json` and `Startup.cs`. If you dont have etiher of these... start again!

Okay so first lets open up `project.json`.

Within the ConfigureServices method add these lines

```c#
services.AddModuleServices(configure => configure
    .AddConfiguration(Configuration)
);
```

Next, at the end of the Configure method, add these lines

```c#
app.UseModules();
```

Thats it. Erm, wait, what? Okay so right now you must be thinking, well what the hell does this do? good question.

AddModuleServices will add the container middleware to you application pipeline, this means in short that It will look in to a folder called Modules within you application for all folders that contain the mainfest file `Module.txt`. IF you looked on the file system it would look like this:

```
MyNewWebApplicaion
  \ Modules
    \ Module1
    \ Module2
```

Once it has found that manifest file, and said file is valid, it will then look for all classes that implement `IStartup`, instansiate them and then call the methods on here. An example of one is using the abstract class off of IStartup

```c#
public class Startup : StartupBase
{
    public override void ConfigureServices(IServiceCollection services)
    {
        services.AddScoped<ISomeInterface, SomeImplementedClass>();
    }
}
```

By doing this you allow your modules to be self contained, completely decoupled from the Hosting applicaiton.

> Note: If you drop a new module in, then you will need to restart the application for it to be found.

## Add Extra Locations
By default module discovery is linked to the `Modules` folder. This however can be extended.

Within the `Startup.cs` file in your host, within the `ConfigureServices` method, add

```c#
services.AddExtensionLocation("SomeOtherFolderToLookIn");
```

## Add Extra Manifest filenames
By default the module manifest file is `Module.txt`. This however can be extended.

Within the `Startup.cs` file in your host, within the `ConfigureServices` method, add

```c#
services.AddManifestDefinition("ManifestMe.txt", "module");
```

## Additional framework
You can add your favourite application framework to the pipeline, easily. The below implementations are designed to work side by side, so if you want Asp.Net Mvc and Nancy within your pipeline, just add both.

The modular framework wrappers below are designed to work directly with the modular application framework, so avoid just adding the raw framework and expect it to just work.

### Asp.Net Mvc
Within your hosting application add a reference to `Microsoft.AspNetCore.Mvc.Modules`

Next, within `Startup.cs` modify the method `AddModuleServices` to look like this

```c#
services.AddModuleServices(configure => configure
    .AddMvcModules(services.BuildServiceProvider())
    .AddConfiguration(Configuration)
);
```

> Note the addition of `.AddMvcModules(services.BuildServiceProvider())`

Thats it, done. Asp.Net Mvc is now part of your pipeline

### NancyFx
Within your hosting application add a reference to `Microsoft.AspNetCore.Nancy.Modules`

Next, within `Startup.cs` modify the method `Configure` to look like this

```c#
app.UseModules(modules => modules
    .UseNancyModules()
);
```

> Note the addition of `.UseNancyModules()`

Thats it, done. NancyFx is now part of your pipeline. What this means is that Nancy modules will be automatically discovered.

> Note. There is no need to register a Nancy Module within its own Startup class.


# Change Notes for Orchard Team

I have moved asp.net mvc out of the Host project Orchard.Cms.Web, and MVC is now a dependency of the Startup project and is also inside of the recipe.

Created a NancyFx implementation, thus allowing us to have Nancy in a module or referenced from the host, similar implementation to MVC.

Removed MVC from the core Orchard startup pipeline

sebastienros commented:
> that inherit off of `StartupBase`

It should use `IStartup` instead, or is that a typo?

How did you do to have custom MVC/Razor configuration done from a module?
- How is the `RouteCollection` handed over to a module, or the MVC configuration object?

I will do some local perf tests to ensure the raw perf has not been impacted (theorically it should have no impact) and the multi-tenant scenario is still good (I expect some impact on this one as you told me the MVC pipeline is different now for each pipeline).

Maybe @jersiovic should try it too as the way MVC is setup might impact his branch.

Jetski5822 commented:
Typo - good catch! - Will answer the other questions on my way home.

sebastienros commented on src/Microsoft.AspNetCore.Mvc.Modules/Extensions/ServiceCollectionBuilderExtensions.cs:
```diff
@@ -67,6 +69,7 @@ public static IMvcCoreBuilder AddExtensionsApplicationParts(this IMvcCoreBuilder
             using (logger.BeginScope("Loading extensions"))
             {
+                ConcurrentBag<Assembly> bagOfAssemblies = new ConcurrentBag<Assembly>();
                 Parallel.ForEach(availableExtensions, new ParallelOptions { MaxDegreeOfParallelism = 4 }, ae =>
```
Why parallel here, it's just enumerating a collection.

jtkech replied:
Not sure in this context but if we still load extensions as we currently do. In the loop `LoadExtensionAsync()` is called which triggers dyn compilation and / or loading.

sebastienros replied:
Are you saying `GetExtensions()` is not actually loading them, but only when we enumerate its result?

jtkech replied:
Yes, GetExtensions() only grab extensions infos based on the search paths options.

jtkech commented on src/Orchard.Environment.Extensions/ExtensionLibraryService.cs:
```diff
@@ -45,16 +43,21 @@ public class ExtensionLibraryService : IExtensionLibraryService
+                        _loadedAssemblies
+                            .Values
+                            .Select(x => x.Value.Location));
+            }
```
I've seen you now pass metadata references to the razor view engine as `MetadataReferenceFeature` through `ApplicationPartManager.FeatureProviders`. Wow, did you try it, does it work? If so, cool.

So, what i can say is that normally here we exclude assemblies which are ambient (part of the app). This by using an hashset which is initialized with `ApplicationAssemblyNames` (see below), not to add them but to exclude them.

jtkech commented on src/Microsoft.AspNetCore.Mvc.Modules/Mvc/ExtensionMetadataReferenceFeatureProvider.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            using (var stream = File.OpenRead(path))
+            {
+                var moduleMetadata = ModuleMetadata.CreateFromStream(stream, PEStreamOptions.PrefetchMetadata);
```
This not to read the entire image into memory 👍

jtkech commented on src/Microsoft.AspNetCore.Modules/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,87 @@
+                internalServices.AddOptions();
+                internalServices.AddLocalization();
+                internalServices.AddHostCore();
+                internalServices.AddExtensionManagerHost("app_data", "dependencies");
```
App_Data

sebastienros commented on src/Orchard.Environment.Shell.Abstractions/Builders/ShellContext.cs:
```diff
@@ -27,7 +27,7 @@ public class ShellContext : IDisposable
         {
-            return ServiceProvider.GetRequiredService<IServiceScopeFactory>().CreateScope();
+            return ServiceProvider.CreateScope();
         }
```
This seems to be a new extension method from 1.1 that does exactly what I wrote. So I assume you can remove this method altogether.

jtkech replied:
Exactly, but maybe we still need the method to use the `ServiceProvider` of a given `ShellContext`.

jtkech replied:
Oh, yes you're right, we could remove it, or keep it as an helper just to write `shellContext.CreateScope()` in place of `shellContext.ServiceProvider.CreateScope()`.

jtkech commented on src/Orchard.Environment.Shell/Builders/ShellContainerFactory.cs:
```diff
@@ -87,20 +86,24 @@ public IServiceProvider CreateContainer(ShellSettings settings, ShellBlueprint b
-                featureServiceCollection.Add(startupServices);
+
                 tenantServiceCollection.Add(startupServices);
+                tenantServiceCollection = new ServiceCollection().Add(tenantServiceCollection.Distinct());
```
Are you sure? E.g about `featureServiceCollection` which has been added to `featureServiceCollections` it seems that here it stays empty. And `featureServiceCollections` is used later to register feature services in `ITypeFeatureProvider`.

Then, `startupServices.Add(featureServiceCollection)` seems to add an empty collection.
Then, maybe you need `startupServices.Add(tenantServiceCollection)` for other reasons.

jtkech replied:
Are you sure? E.g about featureServiceCollection which has been added to featureServiceCollections it seems that here it stays empty. And featureServiceCollections is used later to register feature services in ITypeFeatureProvider.

Then, startupServices.Add(featureServiceCollection) seems to add an empty collection.
Then, maybe you need startupServices.Add(tenantServiceCollection) for other reasons.

jtkech commented on src/Microsoft.AspNetCore.Mvc.Modules/DesignTimeMvcBuilderConfiguration.cs:
```diff
@@ -19,19 +21,14 @@ public void ConfigureMvc(IMvcBuilder builder)
-                ((List<MetadataReference>)options.AdditionalCompilationReferences).AddRange(extensionLibraryService.MetadataReferences());
+                var extensionLibraryService = serviceProvider.GetRequiredService<IExtensionLibraryService>();
+                apm.FeatureProviders.Add(
+                    new ExtensionMetadataReferenceFeatureProvider(extensionLibraryService.MetadataPaths.ToArray()));
```
Anyway here there is a pending PR on `DesignTimeMvcBuilderConfiguration` to be able to resolve shape tag helpers and where we configure services through orchard ... So, we will see ;)

jtkech replied:
Anyway here there is a pending PR on DesignTimeMvcBuilderConfiguration to be able to resolve shape tag helpers and where we configure services through orchard ... So, we will see ;)

jtkech commented:
@Jetski5822 

When i was saying that for `MetadataPaths` we need to exclude ambient assemblies, maybe it is not necessary because just seen that before adding an assembly in `_loadedAssemblies` or `_compileOnlyAssemblies` we always use before `IsAmbientAssembly()`. So maybe i was wrong, sorry.

That said, as i remember, we need `preservecompilationcontext` because we use in some places the default `DependencyContext`. Maybe it could work if you still have `Orchard.Cms.Web.deps.json` in your bin folder.

Best.

Jetski5822 commented:
Yo JT - could do with your help on the compilation side of things.

I'm trying to stay clear of preservecompilationcontext as it places certain constraints on the environment you roll out to.

Sent from my iPhone

> On 28 Jan 2017, at 00:36, jtkech <notifications@github.com> wrote:
> 
> preservecompilationcontext

jtkech commented:
No problem i will search on this but i would need more infos on the constraints it places.

jtkech commented:
@Jetski5822 just done some quick tests but only through `Orchard.Cms.Web`. Here my findings.

- So ok, without `preserveCompilationContext` there is still a `Orchard.Cms.Web.deps.json` file (not tried with the cli) and we can still use e.g `DependencyContext.Default.RuntimeLibraries`. The only difference i've seen is that when the option is used i can get the following.

    ![debug](https://cloud.githubusercontent.com/assets/8586360/22411577/a9b28d7e-e6a5-11e6-8db9-d0c8e387f217.png)

    But without `preserveCompilationContext` the above collection is empty. We use it to compile modules in the same configuration (`Release` / `Debug`) as the ambient app has been built. E.g, the compiler uses it to resolve the binaries paths of referenced projects. Note: When empty we fallback to `Default`.

    That said, if we can't keep this option, here i think we can find another strategy.

- Then, i've seen that modules are not compiled / loaded in parallel. I think it's because now there are 1st loaded through `LoadFeaturesAsync()` when creating the shell. And all `LoadExtensionAsync()` in a `Parallel.ForEach()` are now called after. So, maybe just use a parallel loop in `LoadFeaturesAsync()`.

    Note: we use `MaxDegreeOfParallelism = 4`, maybe we can increase it, VStudio uses 8 by default.

- Also seen that it creates an `app_data` folder with a lower case

- Then, i started `Orchard.Cms.Web` and i got the following.

    ![issue1](https://cloud.githubusercontent.com/assets/8586360/22412598/82f6b1aa-e6b0-11e6-9802-6069c212702b.png)

- So, i tried a fresh setup, then i got the following.

    ![issue2](https://cloud.githubusercontent.com/assets/8586360/22412634/d592c6b0-e6b0-11e6-98b1-37dfb7be7d84.png)

- So, i've tried before your last `cleanup` commit, then i could run the setup view. But i couldn't submit the setup form. Notice the warning message i had on the 1st view rendering before having input anything.

    ![issue3](https://cloud.githubusercontent.com/assets/8586360/22412757/1d9003aa-e6b2-11e6-911f-f7e08af99b80.png)

Best.

Jetski5822 commented:
@jtkech thanks man for the feedback.

The site loads when I add builder.AddTagHelpersAsServices(); - but you end up with it being in a wierd state.

I've removed the Parallel.ForEach(), and will add it to the extension manager in a later PR when I perf test it.

I have also fixed the app_data casing issue.

If you pull latest you will only see the last error you got, which is to do with missing tag helpers.... I am not sure what is going on to be fair. Even with Compilation Context set to true we would still end up with this issue.

I was trying to avoid preserveCompilationContext as we provide the assemblies at run time. preserveCompilationContext is used for the compilation of razor view, but by providing the assemblies at runtime I have worked around that.

Im just not understanding why the tag helper stuff is being such a pain in the arse at the moment... I am wondering if I am providing something that overrides some other implementation somewhere... I just dont know.

What do you think?

Jetski5822 commented on src/Orchard.Environment.Extensions/Features/TypeFeatureProvider.cs:
```diff
@@ -18,7 +21,8 @@ public IFeatureInfo GetFeatureForDependency(Type dependency)
-            throw new InvalidOperationException($"Could not resolve feature for type {dependency.Name}");
+            return CoreFeature;
+            //throw new InvalidOperationException($"Could not resolve feature for type {dependency.Name}");
         }
```
Fixed!

jtkech commented:
@Jetski5822 

**preserveCompilationContext**

- Just re-tried, in fact without this option a `.deps.json` is still created but with less infos.

>From the cli repo: ... **.deps.json contains data used for runtime compilation** ... without this setting, the **compilationOptions will not be present**, and it **will contain only the runtime targets**.

- So, no compilation options and only runtime assemblies. Notice that **compilation assemblies are not always the same as runtime assemblies**. And compile only assembly are not loaded.

- We just use the main app compilation options to use the same config (`Debug` / `Release`) to compile modules. Then we resolve ourselves modules compilation assemblies through the project model. 

- But razor compilation also need compilation assemblies retrieved through `AssemblyPart` which uses `DependencyContext`, [see here](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs#L47).

**Test Orchard.Cms.Web**

- Then, just re-tried with your last commit and, as you, i got the same `TagHelper` issue. Then, i added `AddTagHelpersAsServices()` but i still had the weird issue where i can't input anything.

- After a quick look, i've seen that now some mvc configurations are done through the `Orchard.Mvc` module. So it seems that **some mvc configurations are done in the context of a child container**.

- Normally `ApplicationParts` are already populated with ambient assemblies (our ambient projects, mvc libraries). Then the razor engine retrieve reference from these application parts, [see here](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs#L47).

- So, i think this is why you now need to retrieve all ambient assemblies. But here, you would need to use `DependencyContext` (as razor) to retrieve compilation assemblies (not always the same as runtime ones).

- Indeed, by moving some configurations to the main app, i could run the setup.

**ExtensionLibraryService**

- If we re-use `GetMetadataReferences()` (moved to `MetadataPaths` and not used here), normally we need not to add a `_loadedAssemblies` if it has its counterpart in `_compileOnlyAssemblies`. So, maybe better to use `Union()` in place of `Concat()`.

- Note: Here a Lazy initialization was used before because it was called by the razor engine on each request, so to be done only once. But it seems that now this is no more the case.

Best.

Jetski5822 commented:
@jtkech thanks man, pushed some updates.

I have added a ModularApplicationPart, this should provide all Paths of all assemblies at run time. This is scoped to the request, so that we know we are not serving up modules that are not enabled, which is good.

I have cleaned up the ServiceCollectionExtensions, and added the metadata stuff back in there. That however is no scoped so will need to think about this.

*ExtensionLibraryService:* reverted the metadata to how you were doing it.

Still hitting that dam TagManager thing. I might have to debug the TagManager stuff.... I think its a bug in their stuff as controllers work....

Jetski5822 commented:
I think it might be their code... https://github.com/aspnet/Razor/blob/bdbb854bdbde260b3c70f565a93ebbb185a7c5a7/src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperTypeResolver.cs#L72

This means that it will try to resolve a reference from the Host and not from the module... Ill override it

sebastienros commented on src/Microsoft.AspNetCore.Mvc.Modules/Mvc/ModularApplicationPart.cs:
```diff
@@ -136,34 +138,12 @@ private static IList<Assembly> GetAssemblies(HashSet<string> assemblyNames, Asse
         }
 
-        public static IEnumerable<Assembly> GetModularAssemblies(IServiceProvider services)
+        public static IEnumerable<TypeInfo> GetModularAssemblies(IServiceProvider services)
```
Maybe rename the method?

Jetski5822 commented:
Totally!

Sent from my iPhone

> On 2 Feb 2017, at 21:03, Sébastien Ros <notifications@github.com> wrote:
> 
> @sebastienros commented on this pull request.
> 
> In src/Microsoft.AspNetCore.Mvc.Modules/Mvc/ModularApplicationPart.cs:
> 
> > @@ -136,34 +138,12 @@ private static IList<Assembly> GetAssemblies(HashSet<string> assemblyNames, Asse
>              return locations;
>          }
>  
> -        public static IEnumerable<Assembly> GetModularAssemblies(IServiceProvider services)
> +        public static IEnumerable<TypeInfo> GetModularAssemblies(IServiceProvider services)
> Maybe rename the method?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sebastienros commented on src/Orchard.Cms.Web/Modules/Orchard.Admin/Startup.cs:
```diff
@@ -16,9 +17,13 @@ public override void ConfigureServices(IServiceCollection services)
-            services.AddScoped<IFilterMetadata, AdminZoneFilter>();
-            services.AddScoped<IFilterMetadata, AdminFilter>();
-            services.AddScoped<IFilterMetadata, AdminMenuFilter>();
+            services.Configure<MvcOptions>((options) =>
```
beautiful

Jetski5822 replied:
Merci!

agriffard commented on src/Microsoft.AspNetCore.Mvc.Modules/LocationExpander/ModulerViewLocationExpanderProvider.cs:
```diff
@@ -1,15 +1,14 @@
+namespace Microsoft.AspNetCore.Mvc.Modules.LocationExpander
 {
-    public class ModuleViewLocationExpanderProvider : IViewLocationExpanderProvider
+    public class ModulerViewLocationExpanderProvider : IViewLocationExpanderProvider
```
moduler or modular ?

Jetski5822 replied:
whoops :)

Jetski5822 commented:
Lots of changes today.

 - Removed custom exceptions and are now using just the raw Exception class.
 - Updated nancy module, now its super slick.
 - created a IModularAssemblyProvider to get all assemblies for a given shell. (name might change)
 - Lots of renaming and moving things.

Next up..

 - Remove the Microsoft.AspNetCore.Mvc.Modules.Abstractions project.
 - move HomePageRoute to its own 'thing'
 - Introduce a Swagger module

Jetski5822 commented:
@jtkech check it out and give me your thoughts.

jtkech commented:
@Jetski5822 Sorry for the delay because of other works, just done some tests and here my findings.

**Awesome work**, i like the idea of the modular app part which is a types and references provider.

So, because you add mvc services through a module, we need to provide more types and references  which are related to the main app, this through app parts or razor options for references.

**ModularApplicationPart.Types**

Called many times, each call takes 10ms, 20000 types are returned. In fact, here we can 1st consider only extensions but we also need to add dependencies we now load ourselves. So, i've exposed our loaded assemblies and by only doing the following it works, each call takes 1ms and 9000 types are returned.

    public IEnumerable<TypeInfo> Types
    ...
      var serviceProvider = HttpContextAccessor.HttpContext.RequestServices;
      var extensionLibraryServices = serviceProvider.GetRequiredService<IExtensionLibraryService>();
      return extensionLibraryServices.LoadedAssemblies.SelectMany(x => x.DefinedTypes);

**ModularApplicationPart.GetReferencePaths()**

Called only once and when done here we don't need to do it through razor options. But we can't do it through a regular app part because it would use the dependency context on dynamic loaded modules.

I've seen you use `GetReferencedAssemblies()` but it is related to loaded runtime assemblies, not compilation ones (which may not be loaded). It can works but you can have e.g the same compilation assembly but different runtime ones according to the environment. By using compilation assemblies (see below) i can see that some of them are referenced in the `/refs` subfolder where this kind of assemblies are stored.

So, here we can still provide our `MetadataPaths` but we also need to provide **compilation** references related to the main app, so to the default dependency context. So, by only doing the following it works. Note: 1st it was not working because i didn't see that `preserveCompilationContext` on `Orchard.Cms.Web` was removed again. We need this option to be consistent with runtime compilation, see the last [comment here](https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPart.cs#L47).

    public IEnumerable<string> GetReferencePaths()
    ...
    var serviceProvider = HttpContextAccessor.HttpContext.RequestServices;
    var extensionLibraryServices = serviceProvider.GetRequiredService<IExtensionLibraryService>();

    var assemblyLocations = new List<string>();
    assemblyLocations.AddRange(DependencyContext.Default.CompileLibraries.SelectMany(library => library.ResolveReferencePaths()));
    var locations = assemblyLocations.Union(extensionLibraryServices.MetadataPaths);

    return locations;

**Menu**

The menu shape implemented through a `ShapeTagHelper` was not working. It's because services need to be added in the right order, the `Orchard.Mvc` needs to be called first. I could fix this issue by using a negative priority to this module (higher priorities are called after).

**Little things**

Unused methods: `AddFeatureProvider()`.
Not sure but in my tests with the above code it also works without `AddDefaultFrameworkParts()`.
If references paths are provided through an app part, no need to do it through razor options.

**Dynamic compilation / loading / storing**

Here, just to be aware of what we would need because, beyond dynamic compilation, we are very tied to what dynamic loading and storing do specifically.

See all assemblies in `Orchard.Mvc` bin folder and there is a `/refs` folder for compilation ones as dotnet build do, and see how this folder is stored in `/dependencies`. Note: There could be also a `/runtimes` folder with different runtime ids subfolders as dotnet publish do to store specific runtime assets.

You've done an amazing work and, while we could reconsider what we really need about all the dynamic stuff, you are defining part of the framework dynamically through a module. Just a little afraid :)

Best.

Jetski5822 commented:
hey @jtkech did you pull latest when you took a look?

jtkech commented:
No, sorry forgot to mention that i was still on the `cleanup and removing things that arent needed` commit.

This because i mainly worked on the ModularApplicationPart Types and GetReferencePaths() and i thought that you are using the same here.

So, it works but i'm trying to well understand what happens here and then just add what was missing before. E.g for references, just also add those related to the main app and by using the dependency context to grab compilation assemblies and to be consistent with runtime compilation.

For Types i just re-use the assemblies already loaded when modules was loaded ...

For the menu it's about the main menu on the front end where you have the Home link and which is not displayed. Here we are dependent on the order of module executions. But maybe this now works for you.

jtkech commented:
@Jetski5822 

- Just re-tried with your last commit. So, 1st with the menu it's the same, the front end `main menu` is not displayed (the one which has the `Home` link).

- As mentioned above **i can fix it by using a negative priority** to the `Orchard.Mvc` module, this to force its startup execution to run first. **But this was breaking the setup**.

- This because then during a `Setup` the shape table provider is registered. Then shape tag helpers use shape table provider which uses shell features ... Then, `ShellDescriptorManager` fails because its `_session` is null.

- **I could fix it** by adding a null check, then returning an empty `ShellDescriptor`.

            if (_shellDescriptor == null && _session != null)
            {
                _shellDescriptor = await _session.QueryAsync<ShellDescriptor>().FirstOrDefault();
            }

            return _shellDescriptor ?? new ShellDescriptor();

- For Types and References i'll retry with this last commit and update my comments.

Best.

Jetski5822 commented:
@jtkech hey man, im having issues trying to repo, have you done a `git clean -fxd`?

jtkech commented:
Yes.

But notice that the 1st issue was that the main menu (with the Home link) is not rendered.
So, the work around was to use a negative priority to the  Orchard.Mvc module (then it runs 1st).
Then, when doing this, you get the above issue on a new setup.

Best.

Jetski5822 commented:
Ah okay, im following. Let me have a think about this....

jtkech commented:
So, step by step, here just about app part `Types`.

I've seen that the collection of about 20000 types is now cached. In my above solution where i only use dynamic loaded assemblies i have around 10000 types. Then, just for testing i could get a solution with only 2000 types. So, even it is cached maybe good if we can provide only needed types.

So, i'm learning how mvc 1st add app parts when doing `AddMvcCore()`, see [here ](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/MvcCoreServiceCollectionExtensions.cs#L71). It seems they add libraries which reference some of the mvc assemblies and without including these last ones, see [here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs#L35).

 Note that it needs to resolve an `IHostingEnvironment` which is null in our case. So, for testing i got the host from app services, then registered it before calling `AddMvcCore()`, then no part is added. But if i add in `Orchard.Cms.Web` a reference to mvc, then i get the following where we can see that it now adds a part.

![partmanager](https://cloud.githubusercontent.com/assets/8586360/22623038/5fddeda4-eb4d-11e6-8b4a-711227ec6276.png)

So, just talking about `Types`, because here the main app is mvcless, we don't need to provide types coming through the main app, we only need to consider dyn loaded assemblies which is my first solution with 10000 types. But i think we don't need all.

So, for testing i've only use orchard projects main assemblies, here not only modules but also core projects which are not ambient but referenced by modules. Then it works with only 1860 types.

But maybe a problem if a project reference a library which reference an mvc one, then we would need to include the types of this referenced library. But note that in the master branch currently we don't check this when the main assembly of a module is added to an app part.

Need more thinking and will let you know my findings.

Note: About `GetReferencePaths()` and the needed references of compilation assemblies (not the loaded ones) i'm more clear, i'll post a comment tomorrow.

Best.

jtkech commented:
While testing the above solution with only 1860 types (seems to work until now), due to some recent changes i found another little issue.

**Repro**: In the admin edit the main menu content item, the view fails.

To fix it, in `MenuPart.Edit.cshtml` and `MenuItem.Admin.cshtml` you need to replace.

    Orchard.DisplayManagement.ModelBinding.IModelUpdaterAccessor

To

    Orchard.DisplayManagement.ModelBinding.IUpdateModelAccessor

jtkech commented:
@Jetski5822 about `Types` here my last findings.

Note: I consider separately `IApplicationPartTypeProvider` and `ICompilationReferencesProvider`. 

So, just try to understand, maybe i'm wrong but because i could get it working with about 1800 types, maybe worth to try it, this to provide only 1800 types in place of 20000 types and for each tenant.

- I've done another test where i call `AddMvcCore()` on app services. Then an app part is added only if, in one of the ambient projects, i add a reference on one of the primary mvc assemblies (see below). This confirms that here `Orchard.Cms.Web` is really mvcless, so no need to provide types from the main app.

            // just for testing how mvc adds or not application parts
            var appServices = applicationServices.GetService<IServiceCollection>();
            var appBuilder = appServices.AddMvcCore();

- Then, i've tried to do what mvc do with the main runtime libraries and by using the dependency context, but here only on our dyn loaded assemblies and without the dependency context we can't use for them.

  Here, a comment grabbed [from here](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs#L56).
  >Returns a list of libraries that references the assemblies in <see cref="ReferenceAssemblies"/>.
By default it returns all assemblies that reference any of the primary MVC assemblies
while ignoring MVC assemblies.

  So, i've just copy paste and adapted their `GetCandidateAssemblies()` to return candidate assemblies based on our `LoadedAssemblies` which are our modular runtime assemblies. There are modules main assemblies and their dependencies, unless those which are ambient but we don't care here.

  So, because here we can't use the dependency context, one adaptation, as you have done, is to use recursively `GetReferencedAssemblies()`. We can do this here because we don't deal with compilation references. A 2nd adaptation is to not fail if a dependency is not found because ambient to the app.

  So, i use it like this, without caching it just for testing. Then seems to work with only 1811 types.

        // not cached, only for testing
        public IEnumerable<TypeInfo> Types
        {
            get
            {
                var serviceProvider = HttpContextAccessor.HttpContext.RequestServices;
                var extensionLibraryServices = serviceProvider.GetRequiredService<IExtensionLibraryService>();

                return GetCandidateAssemblies(extensionLibraryServices.LoadedAssemblies)
                    .SelectMany(x => x.DefinedTypes);

                //return AssemblyProvider.GetAssemblies().SelectMany(x => x.DefinedTypes);
            }
        }

Here the implementation of their `GetCandidateAssemblies()` quickly adapted. I'm sure you can make it simpler and, if it is worth to use it, you could integrate it in the right way.

Note: I already seen some uneeded code in our context, e.g no need to check if there are duplicate assemblies in our `LoadedAssemblies`, i will update my post when ready.

        internal static HashSet<string> ReferenceAssemblies { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "Microsoft.AspNetCore.Mvc",
            "Microsoft.AspNetCore.Mvc.Abstractions",
            "Microsoft.AspNetCore.Mvc.ApiExplorer",
            "Microsoft.AspNetCore.Mvc.Core",
            "Microsoft.AspNetCore.Mvc.Cors",
            "Microsoft.AspNetCore.Mvc.DataAnnotations",
            "Microsoft.AspNetCore.Mvc.Formatters.Json",
            "Microsoft.AspNetCore.Mvc.Formatters.Xml",
            "Microsoft.AspNetCore.Mvc.Localization",
            "Microsoft.AspNetCore.Mvc.Razor",
            "Microsoft.AspNetCore.Mvc.Razor.Host",
            "Microsoft.AspNetCore.Mvc.TagHelpers",
            "Microsoft.AspNetCore.Mvc.ViewFeatures"
        };

        // Returns a list of assemblies that references the assemblies in <see cref="ReferenceAssemblies"/>.
        // By default it returns all assemblies that reference any of the primary MVC assemblies
        // while ignoring MVC assemblies.
        // Internal for unit testing
        internal static IEnumerable<Assembly> GetCandidateAssemblies(IEnumerable<Assembly> assemblies)
        {
            if (ReferenceAssemblies == null)
            {
                return Enumerable.Empty<Assembly>();
            }

            var candidatesResolver = new CandidateResolver(assemblies, ReferenceAssemblies);
            return candidatesResolver.GetCandidates();
        }

        private class CandidateResolver
        {
            private readonly IDictionary<string, Dependency> _dependencies;

            public CandidateResolver(IEnumerable<Assembly> assemblies, ISet<string> referenceAssemblies)
            {
                var dependenciesWithNoDuplicates = new Dictionary<string, Dependency>(StringComparer.OrdinalIgnoreCase);
                foreach (var assembly in assemblies)
                {
                    if (!dependenciesWithNoDuplicates.ContainsKey(assembly.GetName().Name))
                    {
                        dependenciesWithNoDuplicates.Add(assembly.GetName().Name, CreateDependency(assembly, referenceAssemblies));
                    }
                }

                _dependencies = dependenciesWithNoDuplicates;
            }

            private Dependency CreateDependency(Assembly assembly, ISet<string> referenceAssemblies)
            {
                var classification = DependencyClassification.Unknown;
                if (referenceAssemblies.Contains(assembly.GetName().Name))
                {
                    classification = DependencyClassification.MvcReference;
                }

                return new Dependency(assembly, classification);
            }

            private DependencyClassification ComputeClassification(string dependency)
            {
                if (!_dependencies.ContainsKey(dependency))
                {
                    return DependencyClassification.Unknown;
                }

                var candidateEntry = _dependencies[dependency];
                if (candidateEntry.Classification != DependencyClassification.Unknown)
                {
                    return candidateEntry.Classification;
                }
                else
                {
                    var classification = DependencyClassification.NotCandidate;

                    foreach (var candidateDependency in candidateEntry.Assembly.GetReferencedAssemblies())
                    {
                        var dependencyClassification = ComputeClassification(candidateDependency.Name);
                        if (dependencyClassification == DependencyClassification.Candidate ||
                            dependencyClassification == DependencyClassification.MvcReference)
                        {
                            classification = DependencyClassification.Candidate;
                            break;
                        }
                    }

                    candidateEntry.Classification = classification;

                    return classification;
                }
            }

            public IEnumerable<Assembly> GetCandidates()
            {
                foreach (var dependency in _dependencies)
                {
                    if (ComputeClassification(dependency.Key) == DependencyClassification.Candidate)
                    {
                        yield return dependency.Value.Assembly;
                    }
                }
            }

            private class Dependency
            {
                public Dependency(Assembly assembly, DependencyClassification classification)
                {
                    Assembly = assembly;
                    Classification = classification;
                }

                public Assembly Assembly { get; }

                public DependencyClassification Classification { get; set; }

                public override string ToString()
                {
                    return $"Library: {Assembly.GetName().Name}, Classification: {Classification}";
                }
            }

            private enum DependencyClassification
            {
                Unknown = 0,
                Candidate = 1,
                NotCandidate = 2,
                MvcReference = 3
            }
        }

Best.

jtkech commented:
Then, about the `ICompilationReferencesProvider` part.

Here again i'm not sure, just thinking it would be better to use the dependency context, as mvc do, when it is possible, here for the main app. For our modules, where it is not possible, we use the project model which will be obsolete without project json files :(

Note: For a precompiled module without projects json files, we know where compile only assemblies are because when we compile modules we store them in a specific subfolder as dotnet do.

So here, in our context, for runtime compilation of razor views we need to add framework references. So here i use the default dependency context to grab `CompileLibraries` related to the main app. Then i do an `Union()` on our `MetadataPaths` related to our modular compilation assemblies.

So, here the goal is to reference compilation assemblies which can be different of runtime ones and then stored in different loacations. As mentioned, to use this solution we need the `preserveCompilationContext` option on the main app. Note: we also use this option to determine under which config we compile modules.

Note: If we add `MetadataPaths` here, we don't need to do it through the razor view engine options.

    var serviceProvider = HttpContextAccessor.HttpContext.RequestServices;
    var extensionLibraryServices = serviceProvider.GetRequiredService<IExtensionLibraryService>();

    return DependencyContext.Default.CompileLibraries
        .SelectMany(library => library.ResolveReferencePaths())
        .Union(extensionLibraryServices.MetadataPaths);

    //return AssemblyProvider.GetAssemblies().Select(x => x.Location);

Best.

jtkech commented:
Just for infos i had an issue when starting a Tenant setup.

Same kind of issue we already fixed but on the main setup, shape tag helpers need shape table manager which needs shell features which needs shell descriptor ... But here the message was saying that `Alpha_Document` is not valide.

I'm not based on your recent commits, so maybe it is already fixed. So, just for infos, here i could fix it by checking the tenant state.

    public async Task<ShellDescriptor> GetShellDescriptorAsync()
    {
        // Prevent multiple queries during the same request
        if (_shellDescriptor == null && _session != null && _shellSettings.State != TenantState.Uninitialized)
        {
            _shellDescriptor = await _session.QueryAsync<ShellDescriptor>().FirstOrDefault();
        }

        return _shellDescriptor ?? new ShellDescriptor();
    }


Best.

Jetski5822 commented:
@jtkech TBH I am a bit confused :). 

Fancy doing a PR to my branch with your ideas?

jtkech commented:
Ok, you're right i will do it, not so easy to be clear because i'm discovering things ;)

So, the idea is to provide **Types** only from assemblies which reference a primary mvc assembly (without including the mvc ones), as mvc do. E.g not provide the 1842 types from System.Private.Corelib. Note: i'm trying to make the above code simpler, e.g without an hardcoded list of primary mvc assemblies names.

About **RefPaths**, the goal is to use the right compilation references by using the dependency context. Here, as a bad example, from the package storage in a dev env, otherwise e.g from the runtime refs subfolder.

    "C:\\Users\\JeanThierry\\.nuget\\packages\\System.Linq\\4.3.0\\ref/netstandard1.6/System.Linq.dll"
    "C:\\Users\\JeanThierry\\.nuget\\packages\\System.Linq.Expressions\\4.3.0\\ref/netstandard1.6/System.Linq.Expressions.dll"

When using GetReferencedAssemblies()

    "C:\\Program Files\\dotnet\\shared\\Microsoft.NETCore.App\\1.1.0\\System.Linq.dll"
    "C:\\Program Files\\dotnet\\shared\\Microsoft.NETCore.App\\1.1.0\\System.Linq.Expressions.dll"


 One of the side effects could be as described in this comment (from an mvc issue) i think is still valid.

> A side effect of this was that the view could be compiling against newer APIs than the app targets (for instance when the app targets net451 but is running using net452)

So, when ready i will rebase and do a PR to show the ideas. Note: i will just consider the mvc part, so maybe it could break what is done e.g for the nancy part.

Best.

Jetski5822 commented:
Okay so woke up with a solution...

At the moment I'm saying that what goes in to the compilation references is the same as what's returned from the modular assembly part. This is wrong.

I will change this! Fail!

Sent from my iPhone

> On 7 Feb 2017, at 22:56, jtkech <notifications@github.com> wrote:
> 
> Ok, you're right i will do it, not so easy to be clear because i'm discovering things ;)
> 
> So, the idea is to provide Types only from assemblies which reference a primary mvc assembly (without including the mvc ones), as mvc do. E.g not provide the 1842 types from System.Private.Corelib. Note: i'm trying to make the above code simpler, e.g without an hardcoded list of primary mvc assemblies names.
> 
> About RefPaths, the goal is to use the right compilation references by using the dependency context. Here, as a bad example, from the package storage in a dev env, otherwise e.g from the runtime refs subfolder.
> 
> "C:\\Users\\JeanThierry\\.nuget\\packages\\System.Linq\\4.3.0\\ref/netstandard1.6/System.Linq.dll"
> "C:\\Users\\JeanThierry\\.nuget\\packages\\System.Linq.Expressions\\4.3.0\\ref/netstandard1.6/System.Linq.Expressions.dll"
> When using GetReferencedAssemblies()
> 
> "C:\\Program Files\\dotnet\\shared\\Microsoft.NETCore.App\\1.1.0\\System.Linq.dll"
> "C:\\Program Files\\dotnet\\shared\\Microsoft.NETCore.App\\1.1.0\\System.Linq.Expressions.dll"
> One of the side effects could be as described in this comment (from an mvc issue) i think is still valid.
> 
> A side effect of this was that the view could be compiling against newer APIs than the app targets (for instance when the app targets net451 but is running using net452)
> 
> So, when ready i will rebase and do a PR to show the ideas. Note: i will just consider the mvc part, so maybe it could break what is done e.g for the nancy part.
> 
> Best.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Jetski5822 commented:
@jtkech Or menu shows, site kicks up. I think there is an issue with the Layouts modules but @sebastienros is going to take a look at that.

The menu is also displaying. I changed how Types are returned too... check it out.

jtkech commented:
@Jetski5822 

I was quite ready to create a PR to show the ideas and all seemed to work on my side. But now i think it's better to 1st do a rebase to see what is done and check it, then maybe create a PR. I will do it.

Meanwhile, after a quick look of some of your recent commits. But need to check it again.

- When i was saying to use **`preserveCompilationContext` it was only for the main app `Orchard.Cms.Web`**, not for modules. This to use the dependency context but only for the main app, not for modules.

Note: without this option we can still use a dependency context for runtime assemblies, but not for compilation ones. We could use this option for modules but it adds many libraries in modules bin folders and as i remember there was some problems with modules assemblies which have not an entry point.

Note: So, currently for modules we use the project model. Notice that there are already more assemblies in modules bin folders because of dependencies of mvc assemblies which are now not ambient.

- So, i use the dependency context only to retrieve compilation references of the main application. For our modules we resolve ourselves compilation references (`MetadataPaths`).

- For Types we need runtime assemblies but here only those that are dynamically loaded because the app is mvcless. And here also we resolve ourselves our modules runtimes assemblies (`LoadedAssemblies`) which are main assemblies of modules and their non ambient dependencies which are dyn loaded.

- So, for Types i use `LoadedAssemblies` **but i have also adapted `GetCandidateAssemblies()` to not use dependency context** because we already provide just needed dynamically loaded dependencies.

But here i need to check it again to see what happens with your last code. Will do it soon.

Best.

Jetski5822 commented:
@sebastienros can you try a bit of perf on this now. 

I still need to look at the amount of types we return for the ModularApplicationPart....

sebastienros commented:
Will try. Also don't forget our discussion about the fact that `ModularApplicationPart` should actually be a `FeatureApplicationPart`.

Jetski5822 commented:
Yeah I will rename

sebastienros commented:
This code:

```
        public IEnumerable<TypeInfo> Types
        {
            get
            {
                return DefaultAssemblyDiscoveryProvider
                    .DiscoverAssemblies(
                        ShellBlueprint.Dependencies.Select(dep => dep.Type.GetTypeInfo().Assembly).ToList(),
                        ReferenceAssemblies)
                    .SelectMany(x => x.DefinedTypes);
            }
        }
```

It's returning all the types of all the dependency assemblies. It should only return the types from the `ShellBlueprint`. That's why you renamed it `ShellFeaturePart`.

Jetski5822 commented:
@sebastienros you think I should have two application parts? one for the shell assemblies, and one for the dependencies?

sebastienros commented:
Maybe as many as there are features, or just one with all features, but they should return only the __TYPES__ that are white listed from the features list, not the full assemblies. You don't want a Controller from a feature that hasn't been enabled. This is why the types from the ShellBluePrint are a very convenient way to do that, they are already filtered.

Jetski5822 commented:
Couldnt I use ITypeFeatureProvider here?

sebastienros commented:
Why can't you return the types of the ShellBluePrint, explain to me what the issue is, looks like I am missing something.

Jetski5822 commented:
You think we should be doing this...;???

```
        public IEnumerable<TypeInfo> Types
        {
            get
            {
                return ShellBlueprint
                    .Dependencies
                    .Select(dep => dep.Type.GetTypeInfo().Assembly)
                    .Distinct()
                    .SelectMany(a => a.DefinedTypes);
            }
        }

        /// <inheritdoc />
        public IEnumerable<string> GetReferencePaths()
        {
            return ShellBlueprint
                .Dependencies
                .Select(dep => dep.Type.GetTypeInfo().Assembly)
                .Distinct()
                .Select(a => a.Location);
        }
```

The issue is the mvc framework needs all the compilation libs too.... its a bit weird. Try it, you will see.

sebastienros commented:
I don't think you get my point, you are still loading all the types of the assemblies.
We already talked about it on skype, i thought it was clear. An extension has an Assembly, but can convey multiple features. Each feature can be enabled independently. MVC should only register the types from the enabled features. Hence you can't return all the types of the assembly when for instance a single feature of this extension is enabled.

The ShellBluePrint contains a collection of types already filtered with only the types associated to the enabled features. 

Here is something that should do it:

```
public IEnumerable<TypeInfo> Types
        {
            get
            {
                return ShellBlueprint
                    .Dependencies
                    .SelectMany(dep => dep.Type);
            }
        }
```

As for the other property (GetReferencePath) I don't know what it represent.

Jetski5822 commented:
Oops, I was running out the door.

That doesn't work as MVC doesn't have the required types to compile.

Will look when I get back, but yeah... it's a bit dumb

Sent from my iPhone

> On 13 Feb 2017, at 17:21, Sébastien Ros <notifications@github.com> wrote:
> 
> I don't think you get my point, you are still loading all the types of the assemblies.
> We already talked about it on skype, i thought it was clear. An extension has an Assembly, but can convey multiple features. Each feature can be enabled independently. MVC should only register the types from the enabled features. Hence you can't return all the types of the assembly when for instance a single feature of this extension is enabled.
> 
> The ShellBluePrint contains a collection of types already filtered with only the types associated to the enabled features.
> 
> Here is something that should do it:
> 
> public IEnumerable<TypeInfo> Types
>         {
>             get
>             {
>                 return ShellBlueprint
>                     .Dependencies
>                     .SelectMany(dep => dep.Type);
>             }
>         }
> As for the other property (GetReferencePath) I don't know what it represent.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sebastienros commented:
maybe you need another ApplicationPArt with something else, and it worked because you just returned everything with this one. What are the missing types?

Also you might keep the same current code for GetReferencePaths (this is for RAzor I think)

jtkech commented:
Here, just for infos the last time i tried nick's branch (not with the latest commits).

**Types**

Before, by using `GetReferencedAssemblies()` about 20000 types was provided, but now, because dependency context doesn't work on a dyn loaded assembly, only 259 Types are provided, only those related to main modules assemblies, not their dependencies, which is not enough.

So, i still use `GetReferencedAssemblies()` on shell features assemblies, but only those which are dyn loaded (app is mvcless) and that reference a primary mvc assembly. Then 1760 Types are provided.

**References**

Here we need the ambient framework compilation assemblies because `AddMvcCore()` is not done at the app level. I think we will not need it if we reference modules and then `Orchard.Mvc` through the main app.

So right now, i use `DependencyContext.Default.CompileLibraries`.

And, for the non ambient libraries, here also we can't use the dependency context, so right now i use our `ExtensionLibraryService.ReferencePaths` which uses the project model (here not filtered by features).

Best.

jtkech commented:
When re-trying my branch, just seen that again the main menu was not always rendered. Not sure it's the case with your branch but i've seen that now the **2 LoadFeaturesAsync()** no longer return ordered features.

So, because of the blueprint features order, even `Orchard.Mvc` has a negative priority, its startup is not always executed first, so shape tag helpers may not be registered ...

Just for infos i could fix it with this [temporary workaround](https://github.com/OrchardCMS/Orchard2/pull/568/commits/e579b09aa6fe210f81a13fe72a6c298eaba0bddb).

Best.

Jetski5822 commented:
@jtkech hang on dude, I'm breaking things at the moment. Use the code before the commit WIP

Sent from my iPhone

> On 14 Feb 2017, at 06:18, jtkech <notifications@github.com> wrote:
> 
> When re-trying my branch, just seen that again the main menu was not always rendered. Not sure it's the case with your branch but i've seen that now the 2 LoadFeaturesAsync() no longer return ordered features.
> 
> So, because of the blueprint features order, even Orchard.Mvc has a negative priority, its startup is not always executed first, so shape tag helpers may not be registered ...
> 
> Just for infos i could fix it with this temporary workaround.
> 
> Best.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sebastienros commented on src/Microsoft.AspNetCore.Mvc.Modules/SatalliteApplicationPart.cs:
```diff
@@ -0,0 +1,114 @@
+        /// Initalizes a new <see cref="AssemblyPart"/> instance.
+        /// </summary>
+        /// <param name="assembly"></param>
+        public SatalliteApplicationPart(IHttpContextAccessor httpContextAccessor)
```
Use the ShellBluePrint directly, or your don't know when it will be resolved

Jetski5822 replied:
I cant, the part gets initialized within a Startup class that doesnt have access to the ShellBlueprint.

sebastienros commented on src/Microsoft.AspNetCore.Mvc.Modules/SatalliteApplicationPart.cs:
```diff
@@ -0,0 +1,114 @@
+    /// <summary>
+    /// An <see cref="ApplicationPart"/> backed by an <see cref="Assembly"/>.
+    /// </summary>
+    public class SatalliteApplicationPart :
```
Typo

Jetski5822 replied:
Fixed

sebastienros commented on src/Microsoft.AspNetCore.Mvc.Modules/SatalliteApplicationPart.cs:
```diff
@@ -0,0 +1,114 @@
+            "Microsoft.AspNetCore.Mvc.ViewFeatures"
+        };
+
+        private bool isInitialized = false;
```
Private members before constructor

Jetski5822 replied:
Fixed

sebastienros commented on src/Microsoft.AspNetCore.Mvc.Modules/SatalliteApplicationPart.cs:
```diff
@@ -0,0 +1,114 @@
+            }
+
+            var shellAssemblies = ShellBlueprint.Dependencies.Keys.Select(type => type.GetTypeInfo().Assembly).ToList();
+            var discoveredAssemblies = DefaultAssemblyDiscoveryProvider
```
Is this code just says that if an assemble has a reference on any of the named assembly from the list, then they are a candidate?

Jetski5822 replied:
Yes, the code was adapted from the MVC project.

jtkech commented:
@Jetski5822 

Just for infos, after all the updates i also had an issue due to bg tasks running during a fresh setup. If you get the same, see here #583 how i fixed it on my side.

Have a good harvest.

Jetski5822 commented:
What!!! @jtkech you not gunna be there?

Sent from my iPhone

> On 18 Feb 2017, at 17:38, jtkech <notifications@github.com> wrote:
> 
> @Jetski5822
> 
> Just for infos, after all the updates i also had an issue due to bg tasks running during a fresh setup. If you get the same, see here #583 how i fixed it on my side.
> 
> Have a good harvest.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jtkech commented:
Not, sorry. Promised, next year ;)

Jetski5822 commented:
@sebastienros Can we now merge this?

jtkech commented:
**I'm ok with the current implementation**, i was too polarized on assemblies deps and types and then not on those related to shell features. So, here only 2 remarks.

-  Just for infos mvc uses `DefinedTypes` and we are using `ExportedTypes`.

- Maybe not important but for infos when we create a tenant container, there are also **DIed** types that we add to `typeFeatureProvider` but which are not added to `shellBluePrint` types.

That said, **if it is still possible to have some missing types** as we seen with shape helpers, **i have another suggestion** which maybe will provide a more complete list of types (without custom code).

The strategy is to 1st grab all mvc types by using the mvc `DefaultAssemblyPartDiscoveryProvider` (not a custom one). Then, we exclude types which are specific to features which are not enabled.

Here, how app mvc types could be grabbed and cached. Then the same list is used by all tenants.

                // we use the public static mvc class which filter mvc types

                _applicationTypes = DefaultAssemblyPartDiscoveryProvider
                    .DiscoverAssemblyParts(hostingEnvironment.ApplicationName)
                    .Where(p => p is AssemblyPart)
                    .SelectMany(p => (p as AssemblyPart).Assembly.ExportedTypes);

                //Note: for each part we could use the regular Types property (but more types)

Then, for a given tenant, something like this.

                var excludedTypes = extensionManager
                    .LoadFeaturesAsync().GetAwaiter().GetResult()
                    .Except(shellBluePrint.Dependencies.Values.Distinct())
                    .SelectMany(f => f.ExportedTypes);

                return GetApplicationTypes()
                    .Except(excludedTypes)
                    .Select(type => type.GetTypeInfo());


I will do a PR just as a suggestion.

I hope your harvest presentation was great and that there will be a video.
